### ISSUE TITLE
Demodulate multiple frequencies within sample simultaneously

### ISSUE THREAD

**Issue by CallumA** [gqrx-sdr/gqrx#196]
I would find it very useful when listening and decoding packet data in particular to be able to demodulate multiple frequencies at once and send them to different UDP ports. This would make it possible to watch multiple streams of packet data at once in real time. It is also useful since there is no ability to record the whole sample (rather than just the demodulated output) so it is not even possible to go back through data to recover another stream.

Also, whilst not specifically needed, it would be nice to have the ability to then quickly switch between selected frequencies while they are demodulated for audio output and maybe have separate frequencies through each channel.

It would be good for when the bookmarks branch is merged too since you could bookmark frequencies to demodulate at once in there.

It should be fine to run multiple demodulations on modern hardware, right now the whole of gqrx is using 50% of a single core (200% being full usage of two cores) on my machine which only has a low power Intel CPU model.

**Comment by zeronullity** [comment]
I agree.. I'm currently editing rtl_tcp.c  source to accomplish this same goal by using multiple gqrx clients & locking the center frequency. It would be nice if this feature was included in gqrx.  I use it mainly for piping udp streams. And being able to pipe multiple udp streams using nearby frequencies in the same bandwidth on a single device is a nice feature. I/Q recorder is not a real-time solution. Trying to use GRC is not efficient and using gqrx-scan is also not a solution.

**Comment by khaytsus** [comment]
Hey, someone is using gqrx-scan :dancers:

**Comment by zeronullity** [comment]
Looks like I may try to make a patch for gqrx,  tcp method will eat up too much CPU resources. Even with 16+ cores it's not a viable method for more than 1 or 2 devices at a time with multiple clients since you would need 1 client per frequency.

**Comment by craigshelley** [comment]
A quick and dirty hack for allowing multiple instances of GQRX on one RTL-SDR
This can be used to demodulate multiple stations simultaneously in real time.
This method involves creating multiple FIFOs, and splitting the raw IQ data to the multiple instances of GQRX
This example is for 2 instances of GQRX
1. Decide on a centre frequency and sample rate that gives the desired band coverage
2. Create 4 FIFOs e.g.
mkfifo /tmp/rawiq0
mkfifo /tmp/rawiq0.iq
mkfifo /tmp/rawiq0a.iq
mkfifo /tmp/rawiq0b.iq
3. Start rtl_sdr to capture raw IQ data (ensure nothing is using the device first)
rtl_sdr -f 152802000 -s 1500000 -g 20 /tmp/rawiq0
4. In a new terminal start sox to convert the 8 bit data to floating point
sox -r 1024 -e unsigned-integer -t raw -b 8 -c 2 /tmp/rawiq0 -t raw -r 1024 -e float /tmp/rawiq0.iq
5. In a new terminal start tee to split the raw floating point streams
tee /tmp/rawiq0a.iq < /tmp/rawiq0.iq > /tmp/rawiq0b.iq
6. Start 2 instances of GQRX with IO device string
file=/tmp/rawiq0a.iq,freq=152802000,rate=1500000.014901,repeat=false,throttle=false
file=/tmp/rawiq0b.iq,freq=152802000,rate=1500000.014901,repeat=false,throttle=false
Ensure the sample rate is configured correctly. In this case 1500000
7 Press the Start DSP Processing button in both instances.

This is a quick and dirty hack, and can probably be simplified by piping the data instead of using FIFOs.

Image shows 2 instances of GQRX on 1 RTL-SDR with 2 instances of multimon-ng decoding the audio streams via UDP.
![gqrx_x2](https://cloud.githubusercontent.com/assets/12436160/7824775/b9a5ccf0-03fd-11e5-9db6-2a3f024b4f84.png)

**Comment by khaytsus** [comment]
Has there been any more consideration here?  GQRX is getting quite powerful, but still doesn't have the native ability to demodulate multiple frequencies at once within the SDR bandwidth.

**Comment by csete** [comment]
Hi @khaytsus 

Sorry to disappoint you but this feature has never been on my own personal road-map, nor I am I aware of anybody else working on it. This is hobby project for me (one of many) and there is too much to do and too little time. So, I have to prioritize according to my own need.

**Comment by milljm** [comment]
A tad bit off topic, but would the following help you out? I ask the question as much as to remind myself to try this when I get home from work:

```pretext
rtl_fm -F1 -o4 -g26 -f 152M:153M:154M -l 50 -s 48000 - | dsd -q -mc -f1 -o pa:1 -i -
```
I recently found that rtl_fm will demodulate several frequencies simultaneously (According to the man page). From what I do know, a single frequency provided to rtl_fm, will work with the above pipe to dsd (done this before successfully)

However, I imagine two broadcasts at the same time will no doubt break the decoding ability of `dsd` as they happen. But might otherwise work when only one broadcast on one of those multiple frequencies being scanned is commencing. Need to test this. I'll edit/delete my post if I find I am completely incorrect... (again, at work, and unable to verify this)

**Comment by khaytsus** [comment]
@milljm CubicSDR can also do this, at the expense of stability and usability (the UI is fairly rough) but it'll demod N frequencies at once..  I don't know how many it'll do, but I've set up 20+ in a given range, although typically only one or two active at any given time.    

I was thinking what you describe is just scanning functionality, not multiple demod?  But you might have a newer rtl_fm than I; I'm not seeing any -o option in mine.

**Comment by milljm** [comment]
@khaytsus I used the term 'demodulate' incorrectly... I am very new to all this :)
I'll give CubicSDR a try.. thanks for that! 

Cheers

**Comment by milljm** [comment]
Okay, I am home, and this is indeed working for me. I am listening to the digital P25 EICAWIN frequencies in my area according to: http://www.radioreference.com/apps/db/?sid=5764

And my successful command is:

```bash
./rtl_fm -F1 -o4 -g26 -f 769207500 -f 769908100 -f 770320700 -f 770432800 -l 50 -s48000 - | dsd -z -mc -f1 -o pa:1 -i -
```
I use the -z option for `dsd` to visualize just how often I get a successful decode... and its not bad at all.
I am using the rtf_fm binary supplied by the Gqrx application (for Mac)

Hope this helps someone down the road. Cheers.

**Comment by kantooon** [comment]
It should be possible to decode multiple digital streams using gr-dsd. In fact I have already integrated gr-dsd with Gqrx and tested this. Of course to have multiple streams extensive changes to the codebase of Gqrx would be required.

On 5 December 2016 21:22:48 GMT, Jason Miller <notifications@github.com> wrote:
>A tad bit off topic, but would the following help you out? I ask the
>question as much as to remind myself to try this when I get home from
>work:
>
>```pretext
>rtl_fm -F1 -o4 -g26 -f 152M:153M:154M -l 50 -s 48000 - | dsd -q -mc -f1
>-o pa:1 -i -
>```
>I recently found that rtl_fm will demodulate several frequencies
>simultaneously (According to the man page). From what I do know, a
>single frequency provided to rtl_fm, will work with the above pipe to
>dsd (done this before successfully)
>
>However, I imagine two broadcasts at the same time will no doubt break
>the decoding ability of `dsd` as they happen. But might otherwise work
>when only one broadcast on one of those multiple frequencies being
>scanned is commencing. Need to test this. I'll edit/delete my post if I
>find I am completely incorrect... (again, at work, and unable to verify
>this)

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by pe1chl** [comment]
It would be helpful when the I/Q recorder could record to a UDP stream (and for the usecase depicted here, be able to playback from a UDP stream as well).
I have another use case where only recording would be helpful already.
I see in the source that GNU radio unfortunately has two different modules for this, file_sink and udp_sink, and that GQRX explicitly calls file_sink.
So it would require a new option to select UDP instead of a file, and handle that by calling udp_sink instead.
